sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -99,6 +99,13 @@ abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging {
    */
   lazy val resolved: Boolean = expressions.forall(_.resolved) && childrenResolved
 
+  /**
+   * Returns true if the two plans are semantically equal. This should ignore state generated
+   * during planning to help the planning process.
+   * TODO: implement this as a pass that canonicalizes the plan tree instead?
+   */
+  def semanticEquals(other: LogicalPlan): Boolean = this == other
```

Comment:
Oh, this is a new semantic equals.  How is this different than `sameResult`?  Maybe we should unify the naming between Expression and LogicalPlan for this concept.